pkg/providers/snow/reconciler/clientbuilder.go

```diff
@@ -0,0 +1,97 @@
+package reconciler
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+
+	awstypes "github.com/aws/aws-sdk-go-v2/aws"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	awscredentials "github.com/aws/aws-sdk-go-v2/credentials"
+	"gopkg.in/ini.v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/eks-anywhere/pkg/aws"
+)
+
+type awsClientBuilder struct {
+	client client.Client
+}
+
+func NewAwsClientBuilder(client client.Client) *awsClientBuilder {
+	return &awsClientBuilder{
+		client: client,
+	}
+}
+
+func (b *awsClientBuilder) BuildSnowAwsClientMap(ctx context.Context) (aws.Clients, error) {
+	credentials, certificates, err := SnowCredentials(ctx, b.client)
+	if err != nil {
+		return nil, fmt.Errorf("getting snow credentials: %v", err)
+	}
+
+	return createAwsClients(ctx, credentials, certificates)
+}
+
+type credentialConfiguration struct {
+	AccessKey string `ini:"aws_access_key_id"`
+	SecretKey string `ini:"aws_secret_access_key"`
+	Region    string `ini:"region"`
+}
+
+func createAwsClients(ctx context.Context, credentials []byte, certificates []byte) (aws.Clients, error) {
+	var deviceIps []string
+	credsCfg, err := ini.Load(credentials)
+	if err != nil {
+		return nil, fmt.Errorf("loading values from credentials: %v", err)
+	}
+	for _, ip := range credsCfg.SectionStrings() {
+		if net.ParseIP(ip) != nil {
+			deviceIps = append(deviceIps, ip)
+		}
+	}
+
+	deviceClientMap := make(aws.Clients, len(deviceIps))
+	for _, ip := range deviceIps {
+		ipCfg, err := parseIpConfiguration(credsCfg, ip)
+		if err != nil {
+			return nil, fmt.Errorf("parsing configuration for %v: %v", ip, err)
+		}
+		clientCfg, err := awsconfig.LoadDefaultConfig(ctx,
```

Comment:
curious why we use a different config than when building the clients from the cli?

Reply:
We load it from a secret as data instead of files, so the whole logic of creating the client actually changes.

Reply:
Got it. But once we load the data, shouldn't the rest of the logic be the same. In other words, is it possible to separate "data load" from "client construction"? And then reuse the latter in both scenarios

Reply:
It doesn't because of how you load this config from this sdk. I am using static credentials provider instead of a shared credentials provider through file inputs. Here is the comparison: https://github.com/aws/eks-anywhere/blob/0c56cd3dfe592057433c76977e96172afd5438d4/pkg/aws/snow.go#L79